src/fsharp/opt.fs

```diff
@@ -1440,6 +1442,34 @@ let ExpandStructuralBinding cenv expr =
               mkLetsBind m binds (mkLet tgtSeqPtOpt m v tuple body)
               (* REVIEW: other cases - records, explicit lists etc. *)
     | expr -> expr
+
+// Moves outer tuple binding inside near the tupled expression:
+// let t = (let a0=v0 in let a1=v1 in ... in let an=vn in e0,e1,...,em) in body
+// let a0=v0 in let a1=v1 in ... in let an=vn in (let t = e0,e1,...,em in body)
+// This way ExpandStructuralBinding can replace expressions in constants, t is directly bound
+// to a tuple expression so that other optimizations such as OptimizeTupleFieldGet work,
+// and the tuple allocation can be eliminated.
+// Most importantly, this successfully eliminates tuple allocations for implicitly returned
+// formal arguments in method calls.
+let rec RearrangeTupleBindings expr fin =
+    match expr with
+    | Expr.Let (bind,body,m,_) ->
+        match RearrangeTupleBindings body fin with
+        | Some b -> Some (mkLetBind m bind b)
+        | None -> None
+    | Expr.Op (TOp.Tuple,_,_,_) ->
```

Comment:
In principle this optimization can flatten a much broader range of expressions, e.g. ones that end in any TOp, App, Lambda etc.

I don't think I can immediately convince you that we should do this, but the nesting-of-let-bindings is common in the compiler, and we should consider over time whether we extend the applicability of this technique.